### PR TITLE
[c10d][Partial-Graph Overlap] Support calling .wait_tensor() on output tensor of eager `async_op=True` collective if under `allow_inflight_collective_as_graph_input_ctx()` context manager

### PR DESCRIPTION
**Pull request by yf225**
This PR aims to support the following use case:
```python
def all_reduce_eager(x):
    y = x * x
    req = dist.all_reduce(y, op=dist.ReduceOp.SUM, async_op=True)
    assert isinstance(req, torch.distributed.Work)
    return y

@torch.compile(fullgraph=True)
def all_reduce_wait_compiled(y):
    torch.ops.c10d_functional.wait_tensor(y)
    return y * y

x = torch.ones(1280, 1280, device="cuda") + self.rank
with allow_inflight_collective_as_graph_input_ctx():
    y = all_reduce_eager(x)
    z = all_reduce_wait_compiled(y)
```
where the collective is issued in eager (with `async_op=True`) but waited in compiled region.

This is important for internal use cases such as TorchRec, where we issue collectives in eager for SparseArch all_to_all but want to wait for them in compiled region at beginning of OverArch, so that the all_to_all can be overlapped with the DenseArch compute that runs in parallel.

----

**Update**: Did two items to prevent regression to existing use cases:

1. Added memory-stressed test case to test_c10d_nccl.py `test_unwaited` to cover existing user's "not calling work.wait() for non-functional collective" use case
2. Gated all new `register_work()` / `unregister_work()` calls with `c10d::allow_inflight_collective_as_graph_input()` check, which is a new context manager that requires explicit user enablement (i.e. not on by default, so should not affect existing users).

The risk of this new version of PR causing regression should be very low.

------

Test commands:
- `pytest -rA test/distributed/test_inductor_collectives.py::TestCollectivesMultiProc::test_eager_async_allreduce_inductor_wait`
- `pytest -rA test/test_fx.py::TestDCE::test_keep_collectives`
- `pytest -rA test/test_fx.py::TestDCE::test_keep_collectives_no_overload`
- `pytest -rA test/distributed/test_c10d_functional_native.py::TestWithNCCL::test_wait_tensor`
- `pytest -rA test/distributed/test_c10d_functional_native.py::TestWithNCCL::test_unwaited`
- `pytest -rA test/distributed/test_c10d_nccl.py::CommTest::test_wait_tensor`
- `pytest -rA test/distributed/test_c10d_nccl.py::CommTest::test_unwaited`
- `pytest -rA test/distributed/_tensor/test_tensor_ops.py::DistTensorOpsTest::test_equal`
- `pytest -rA test/distributed/_tensor/test_random_ops.py::DistTensorRandomOpTest::test_manual_seed`
- `pytest -rA test/distributed/test_dynamo_distributed.py::TestMultiProc::test_ddp_baseline_aot_eager_multiprocess`
- `pytest -rA test/distributed/test_dynamo_distributed.py::TestMultiProc::test_fsdp_aot_eager`
- `pytest -rA test/distributed/test_dynamo_distributed.py::TestMultiProc::test_fsdp_setattr`
- `pytest -rA test/distributed/test_dynamo_distributed.py::TestMultiProc::test_fsdp_unspecialized_forced_getattr_inline`
- `pytest -rA test/distributed/test_dynamo_distributed.py::TestMultiProc::test_fsdp_unspecialized_forced_getattr_no_inline`
- `pytest -rA test/distributed/test_dynamo_distributed.py::TestMultiProc::test_asymmetric_compilation`
- `pytest -rA test/distributed/test_dynamo_distributed.py::TestMultiProc::test_compiler_collectives_automatic_dynamic_scalar`
- `pytest -rA test/distributed/test_dynamo_distributed.py::TestMultiProc::test_compiler_collectives_automatic_dynamic_speculation_divergence`
- `pytest -rA test/distributed/test_dynamo_distributed.py::TestMultiProc::test_compiler_collectives_automatic_dynamic_tensor`
- `pytest -rA test/distributed/test_dynamo_distributed.py::TestMultiProc::test_compiler_collectives_dim_mismatch`
- `pytest -rA test/distributed/test_dynamo_distributed.py::TestMultiProc::test_compiler_collectives_graph_break_empty_graph_still_collective`
- `pytest -rA test/distributed/test_dynamo_distributed.py::TestMultiProc::test_compiler_collectives_missing_source`
- `pytest -rA test/distributed/test_dynamo_distributed.py::TestMultiProc::test_compiler_collectives_scalar_missing_source`
- `pytest -rA test/distributed/test_dynamo_distributed.py::TestMultiProc::test_compiler_collectives_type_mismatch`
- `pytest -rA test/distributed/test_dynamo_distributed.py::TestMultiProc::test_ddp_activation_checkpointing`
- `pytest -rA test/distributed/test_dynamo_distributed.py::TestMultiProc::test_ddp_baseline_aot_eager_multiprocess`
- `pytest -rA test/distributed/test_dynamo_distributed.py::TestMultiProc::test_fsdp_activation_checkpointing`
- `pytest -rA test/distributed/test_dynamo_distributed.py::TestMultiProc::test_fsdp_aot_eager`
- `pytest -rA test/distributed/test_dynamo_distributed.py::TestMultiProc::test_fsdp_inductor`
- `pytest -rA test/distributed/test_dynamo_distributed.py::TestMultiProc::test_fsdp_setattr`
- `pytest -rA test/distributed/test_dynamo_distributed.py::TestMultiProc::test_fsdp_unspecialized_forced_getattr_inline`
- `pytest -rA test/distributed/test_dynamo_distributed.py::TestMultiProc::test_fsdp_unspecialized_forced_getattr_no_inline`
- `pytest -rA test/distributed/test_dynamo_distributed.py::TestMultiProc::test_hf_bert_ddp_aot_eager`
- `pytest -rA test/distributed/test_dynamo_distributed.py::TestMultiProc::test_hf_bert_ddp_aot_eager_static_graph`
- `pytest -rA test/distributed/test_dynamo_distributed.py::TestMultiProc::test_hf_bert_ddp_inductor`
- `pytest -rA test/distributed/test_dynamo_distributed.py::TestMultiProc::test_hf_bert_ddp_inductor_static_graph`
- `pytest -rA test/distributed/test_dynamo_distributed.py::TestMultiProc::test_hf_bert_fsdp_activation_checkpointing`
- `pytest -rA test/distributed/_tensor/test_experimental_ops.py::DistOtherOpsTest::test_bernoulli`
- `pytest -rA test/distributed/_tensor/test_dtensor_compile.py::TestDTensorCompileE2E::test_tp_compile_fullgraph_is_seq_parallel_True`
- `pytest -rA test/distributed/test_inductor_collectives.py::TestCollectivesMultiProc::test_allreduce_inductor_cudagraph_trees`
- `python benchmarks/dynamo/torchbench.py --ci --accuracy --timing --explain --inductor --device cuda --inference --bfloat16 --total-partitions 2 --partition-id 1 --output inference_torchbench.csv --only moco`

------

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #137763

cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @rec

Differential Revision: [D65023311](https://our.internmc.facebook.com/intern/diff/D65023311)

### CODE REVIEW

**Comment by yifuwang** [comment]
Hmm I think this might be problematic. The registration mechanism is designed for the function collective subsystem in which we know in-flight tensors are meant to be waited by `_c10d_functional.wait_tensor`, which removes the work from the registry on completion. Unconditionally registering work here would cause memory leak, because in non-compiled use cases, we would create a permanent entry for each collective call. It will affect all use cases of ProcessGroupNCCL and these will trigger the "unwaited c10d_functional collective calls" warning.

I think the general idea is sound. One thing though is that a user should never throw away the work associated with an in-flight tensor before waiting on it. So can we just require manual work registration?

```python
def all_reduce_eager(x):
    y = x * x
    req = dist.all_reduce(y, op=dist.ReduceOp.SUM, async_op=True)
    assert isinstance(req, torch.distributed.Work)
    torch._C._distributed_c10d._register_work(y, res)
    return y

@torch.compile(fullgraph=True)
def all_reduce_wait_compiled(y):
    torch.ops.c10d_functional.wait_tensor(y)
    return y * y
```

**Comment by yf225** [comment]
Discussed offline:

1. Keep `WorkRegistry` global, move the code from Functional.cpp into base ProcessGroup.cpp file
2. Add register_work to each collective in ProcessGroup.hpp, to make it general to all backends not just NCCL
3. Remove the registry entry in Work::finish

File: torch/csrc/distributed/c10d/ProcessGroup.cpp [hunk 1]
```diff
@@ -158,3 +159,105 @@ void ProcessGroup::release_resources() {
...
+    }
+  }
+
+  size_t get_work_registry_size() {
+    std::unique_lock lock(lock_);
+    return registry_.size();
+  }
+
+  ~WorkRegistry() {
+    unregister_completed_works();
```
Comment (yf225) [inline torch/csrc/distributed/c10d/ProcessGroup.cpp:207]: Unfortunately there is no reliable callback method where we can insert the "unregister_work()" cleanup call (`Work::finish()` is only called in Gloo PG but not NCCL PG, thus we can't use it). Instead we can do `unregister_completed_works()` here to do the cleanup for completed works before throwing the warning for uncompleted work.

File: torch/csrc/distributed/c10d/ProcessGroup.cpp [hunk 2]
```diff
@@ -115,7 +116,9 @@ ProcessGroup::ProcessGroup(
 ProcessGroup::ProcessGroup(int rank, int size)
     : rank_(rank), size_(size), backendType_(BackendType::UNDEFINED) {}
 
-ProcessGroup::~ProcessGroup() = default;
+ProcessGroup::~ProcessGroup() {
+  c10d::unregister_completed_works();
+}
```
Comment (yf225) [inline torch/csrc/distributed/c10d/ProcessGroup.cpp:121]: Unfortunately there is no reliable callback method where we can insert the "unregister_work()" cleanup call for the pure-eager case (`Work::finish()` is only called in Gloo PG but not NCCL PG, thus we can't rely on it). Instead we can do `unregister_completed_works()` here to do the cleanup for completed works, before throwing the warning in `~WorkRegistry()` for the uncompleted works. `ProcessGroup` instance bound to Python should always be destroyed by Python shutdown sequence before the C++ per-thread static object `WorkRegistry`, so doing the cleanup here should work.

File: torch/csrc/distributed/c10d/ProcessGroup.cpp [hunk 3]
```diff
@@ -158,3 +161,137 @@ void ProcessGroup::release_resources() {
...
+    auto it = registry_.find(storage);
+    if (it == registry_.end()) {
+      return {};
+    }
+    auto works = it->second;
+    registry_.erase(it);
+    return works;
+  }
+
+  void unregister_completed_works() {
```
Comment (yf225) [inline torch/csrc/distributed/c10d/ProcessGroup.cpp:210]: From my experience fixing the CI errors, this function is usually the culprit. In case we think the CI might not be sufficiently covering all production use cases, we could make this feature into a gradual rollout to minimize potential impact.

**Comment by yf225** [comment]
@yf225 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D64511994).

**Comment by yf225** [comment]
@yf225 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D64511994).

File: torch/csrc/distributed/c10d/ProcessGroup.cpp [hunk 4]
```diff
@@ -158,3 +159,149 @@ void ProcessGroup::release_resources() {
...
+
+namespace c10d {
+
+void register_work(
+    const at::Tensor& tensor,
+    const c10::intrusive_ptr<c10d::Work>& work) {
+  // Always clean up previously completed work objects, so that even if
+  // the user keeps issuing new collectives without waiting on previous ones,
+  // the registry size would not grow unbounded.
+  RankLocal<WorkRegistry>::get().unregister_completed_works();
```
Comment (yifuwang) [inline torch/csrc/distributed/c10d/ProcessGroup.cpp:287]: Hmm do we need/want to go through the registry everytime a new work is registered? Also it seems that the cleanup wouldn't happen for the last work entry in the program?

As discussed, I really think the entries should be evicted by `Work::wait()` on a per work basis. Do you think there's a problem with that approach?

File: torch/csrc/distributed/c10d/ProcessGroup.cpp [hunk 5]
```diff
@@ -158,3 +159,149 @@ void ProcessGroup::release_resources() {
...
+    auto storage = tensor.storage().getWeakStorageImpl();
+    std::unique_lock lock(lock_);
+
+    auto it = registry_.find(storage);
+    if (it == registry_.end()) {
+      registry_.emplace(
+          std::move(storage),
+          std::vector<c10::intrusive_ptr<c10d::Work>>{work});
+    } else {
+      // There is no guarantee that the previous work object for this
```
Comment (yifuwang) [inline torch/csrc/distributed/c10d/ProcessGroup.cpp:187]: Interesting. This make sense.

File: torch/csrc/distributed/c10d/ProcessGroup.hpp [hunk 6]
```diff
@@ -23,6 +24,16 @@ constexpr auto kProcessGroupDefaultTimeout =
...
 
+C10_EXPORT void register_work(
+    const at::Tensor& tensor,
+    const c10::intrusive_ptr<c10d::Work>& work);
+
+C10_EXPORT at::Tensor wait_tensor(const at::Tensor& tensor);
+
+C10_EXPORT void unregister_completed_works();
+
+C10_EXPORT size_t get_work_registry_size();
```
Comment (yifuwang) [inline torch/csrc/distributed/c10d/ProcessGroup.hpp:35]: Are these two functions only used for testing purposes?

File: torch/csrc/distributed/c10d/ProcessGroup.cpp (hunk 4 above)
Comment (yifuwang) [reply torch/csrc/distributed/c10d/ProcessGroup.cpp:287]: > As discussed, I really think the entries should be evicted by Work::wait() on a per work basis. Do you think there's a problem with that approach?

After reading more, it seems that we now have a one-to-many relationship from storage -> work. Eviction on a per-work basis would be more complex now, but I still think we should do it, because otherwise the registry is keeping the work alive for longer.

I suggesting something like `WorkRegistry::unregister_work(const c10::intrusive_ptr<c10d::Work>& work)` and call it in `Work::wait()`. We go through the registry and remove the **specified** work from the list associated with every storage. If a list becomes empty, we remove the entry from the registry.

Also there are some potential issues with your current impl: (1) not all work implementation has `checkCompleted` (2) `checkCompleted` can be quite expensive (`WorkNCCL::checkCompleted` does a bunch of things and an event query: https://fburl.com/code/qu5hi37w, doing these for all active work whenever a new work is registered may cause some unwanted effects). The suggestion above should resolve these issues because it only checks the identity of the work.

**Comment by yifuwang** [comment]
The general approach looks good!

One issue is that we shouldn't and don't need to rely on `WorkNCCL::isCompleted` for entry removal. A work can be removed from the registry as long as its waited (i.e. EventWait issued from CPU). However, unless a job is running with `CUDA_LAUNCH_BLOCKING`, a collective is most likely not finished on GPU when its work.wait() is called on CPU. Relying on `WorkNCCL::isCompleted` extends the lifetime of the works in a similar way that `record_stream` extends the lifetime of allocations, which is not ideal.

Fortunately we don't need to rely on it. Please see my other comment for a suggestion.

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp [hunk 7]
```diff
@@ -720,6 +720,9 @@ void ProcessGroupNCCL::WorkNCCL::handleException(
 
 void ProcessGroupNCCL::WorkNCCL::synchronize() {
   synchronizeStream();
+  c10d::unregister_work(
+      c10::intrusive_ptr<
+          ProcessGroupNCCL::WorkNCCL>::unsafe_reclaim_from_nonowning(this));
```
Comment (yf225) [inline torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:725]: seems that both `WorkNCCL::wait()` and `WorkNCCL::synchronize()` are valid "wait" APIs, so putting unregister_work in `synchronize()` since it's also called by `wait()`

File: torch/csrc/distributed/c10d/Work.cpp [hunk 8]
```diff
@@ -70,7 +71,10 @@ std::vector<at::Tensor> Work::result() {
   TORCH_CHECK(false, "result() not implemented.");
 }
 
-void Work::synchronize() {}
+void Work::synchronize() {
+  c10d::unregister_work(
+      c10::intrusive_ptr<Work>::unsafe_reclaim_from_nonowning(this));
+}
```
Comment (yf225) [inline torch/csrc/distributed/c10d/Work.cpp:77]: seems both `wait()` and `synchronize()` should serve the same purpose (similar reasoning as in WorkNCCL)

File: torch/csrc/distributed/c10d/ProcessGroup.hpp (hunk 6 above)
Comment (yf225) [reply torch/csrc/distributed/c10d/ProcessGroup.hpp]: yes - `get_work_registry_size` is only for testing purposes

File: torch/csrc/distributed/c10d/ProcessGroup.cpp (hunk 4 above)
Comment (yf225) [reply torch/csrc/distributed/c10d/ProcessGroup.cpp]: sounds great - changed to call `unregister_work()` in `Work::wait()`

**Review by yifuwang — APPROVED** [review]
Awesome work!

**Comment by yf225** [comment]
Discussed with @yifuwang :
1. For functional collective, we need to retain strong pointer to Work in registry; for normal collective, we can retain weak pointer.
2. Best to have two registries: 1 strong and 1 weak.
3. Double-registration (1 strong + 1 weak) for functional collective (and double-warning for them) is okay.

**Comment by wconstab** [comment]
I had a question based on the example in the PR desc- what does it mean to torch compile a function but mark it as full graph=true? Clearly you are trying to leave the eager ops out of the compiled region which suggests not a full graph.

**Comment by yf225** [comment]
> I had a question based on the example in the PR desc- what does it mean to torch compile a function but mark it as full graph=true? Clearly you are trying to leave the eager ops out of the compiled region which suggests not a full graph.

@wconstab It's only fullgraph within the function (i.e. we will not graph-break on the `.wait_tensor` or the `y * y`)

**Comment by yf225** [comment]
@yf225 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D65023311).

**Comment by yf225** [comment]
Discussed with @yifuwang offline on the API design, will land this version to unblock internal use cases.

**Comment by yf225** [comment]
Did two items to prevent regression:

1. Added memory-stressed test case to test_c10d_nccl.py `test_unwaited` to cover existing user's "not calling work.wait() for non-functional collective" use case
2. Gated all new `register_work()` / `unregister_work()` calls with `c10d::allow_inflight_collective_as_graph_input()` check.

The risk of this new version of PR causing regression should be very low.
